Geometry/Surfaces/QuadSurfaces/cone_class.f90

```diff
@@ -0,0 +1,417 @@
+    aabb(self % plane + 3) = self % vertex(self % plane) + maxRadius * sqrt(self % tan2)
+
+  end function boundingBox
+
+  !!
+  !! Evaluate surface expression c = F(r)
+  !!
+  !! See surface_inter for details
+  !!
+  pure function evaluate(self, r) result(c)
+    class(cone), intent(in)                 :: self
+    real(defReal), dimension(3), intent(in) :: r
+    real(defReal)                           :: c, cMin, cMax, cBasis
+    real(defReal), dimension(3)             :: diff
+
+    ! Vector for the difference between position provided and vertex
+    diff = r - self % vertex
+
+    ! Evaluate the expression for the surface of the cone
+    c = dot_product(diff(self % plane), diff(self % plane)) - &
+        self % tan2 * diff(self % axis) ** 2
+
+    ! Evaluate the expressions for the two basis of the cone and find the maximum
+    cMin = - diff(self % axis) + self % hMin
+    cMax = diff(self % axis) - self % hMax
+    cBasis = max(cMin, cMax)
+
+    ! Find the overall maximum
+    c = max(c, cBasis)
+
+  end function evaluate
+
+  !!
+  !! Return distance to the surface
+  !!
+  !! See surface_inter for details
+  !!
+  !! Solves quadratic intersection equation
+  !!   a*d^2 + 2b*d + c = 0
+  !!   c = cone surface expression
+  !!   b = (r1 - v1)u1 + (r2 - v2)u2 - t^2(r3 - v3)u3
+  !!   a = u1^2 + u2^2 - t^2 * u3^2
+  !!
+  pure function distance(self, r, u) result(dist)
+    class(cone), intent(in)                 :: self
+    real(defReal), dimension(3), intent(in) :: r
+    real(defReal), dimension(3), intent(in) :: u
+    real(defReal)                           :: dist
+    real(defReal), dimension(3)             :: diff
+    real(defReal), dimension(2)             :: testPoint2
+    real(defReal)                           :: a, b, c, delta, d1, d2, testPoint1, &
+                                               testRadius, maxRadius, distBasis
+
+    ! Vector for the difference between position provided and vertex
+    diff = r - self % vertex
+
+    ! Calculate quadratic components in the cone
+    a = dot_product(u(self % plane), u(self % plane)) - &
+        self % tan2 * u(self % axis)**2
+    b = dot_product(diff(self % plane) , u(self % plane)) - &
+        self % tan2 * diff(self % axis) * u(self % axis)
+    c = dot_product(diff(self % plane), diff(self % plane)) - &
+        self % tan2 * diff(self % axis) ** 2
+
+    ! Calculate delta/4
+    delta = b * b - a * c
+
+    ! Calculate the distances from the cone surface
+    if (a == ZERO) then
+      d1 = - HALF * c / b
+      d2 = d1
+
+    elseif (delta < ZERO) then
+      d1 = INF
+      d2 = d1
+
+    else
+      d1 = - (b + sqrt(delta)) / a
+      d2 = - (b - sqrt(delta)) / a
+
+    end if
+
+    ! If one the distance found is smaller than the surface tolerance, then the particles
+    ! is on a surface: set that distance to infinity
+    if (d1 < self % surfTol()) d1 = INF
+    if (d2 < self % surfTol()) d2 = INF
```

Comment:
I am not sure if this approach will work I'm afraid... :-( 
The problem are particles hitting the surface at very shallow angle. They may be well within the tolerance but the distance to the surface may be still quite large. This is why in the e.g. 'cylinder' we need an extra case in the if-else for when a particle is within surface tolerance. 

Sorry... geometry is super annoying.

Reply:
I see, I hadn't thought about that! I will reintroduce the extra if-else case when solving the quadratic, hopefully that will do it!